### PR TITLE
Don't resolve to rectracted package versions

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: lib/src/solver/package_lister.dart [hunk 1]
```diff
@@ -36,6 +36,13 @@ class PackageLister {
   /// solve isn't a `pub get`.
   final PackageId _locked;
 
+  // The version of this package that, if retracted, is still allowed in the
+  // current version solve.
+  //
+  // We don't allow a retracted version during solving unless it is already
+  // present in pubspec.lock or pinned in dependency_overrides.
```
Comment (jonasfj) [inline lib/src/solver/package_lister.dart:43]:
```suggestion
  // present in `pubspec.lock` or pinned in `dependency_overrides`.
```

Comment (jonasfj) [reply lib/src/solver/package_lister.dart:43]: Should this just be a `Version`, or not a `PackageId`? maybe it doesn't matter...

Comment (jonasfj) [reply lib/src/solver/package_lister.dart:43]: Please document that this is `null` if no exceptions are to be made.

File: lib/src/solver/version_solver.dart [hunk 2]
```diff
@@ -431,7 +431,9 @@ class VersionSolver {
       List<PackageId> packages) async {
```
Comment (jonasfj) [inline lib/src/solver/version_solver.dart:431]: Should we specify that `_getAvailableVersions` never includes retracted versions, instead? Where is the result used?

File: lib/src/solver/version_solver.dart [hunk 3]
```diff
@@ -500,6 +508,17 @@ class VersionSolver {
     return _lockFile.packages[package];
   }
 
+  /// Gets the version of [package] which can be allowed during version solving
+  /// even if that version is marked as retracted.
+  ///
+  /// We only allow resolving to a retracted version if it is already in the
+  /// pubspec.lock or pinned in dependency_overrides.
```
Comment (jonasfj) [inline lib/src/solver/version_solver.dart:515]:
```suggestion
  /// `pubspec.lock` or pinned in `dependency_overrides`.
```

File: lib/src/source.dart [hunk 4]
```diff
@@ -161,15 +161,20 @@ abstract class BoundSource {
   /// Sources should not override this. Instead, they implement [doGetVersions].
   ///
   /// If [maxAge] is given answers can be taken from cache - up to that age old.
-  Future<List<PackageId>> getVersions(PackageRef ref, {Duration maxAge}) {
+  ///
+  /// The [allowedRetractedVersion], if given, is the only version which can be
+  /// selected even if it is marked as retracted. Otherwise, all the returned
```
Comment (jonasfj) [inline lib/src/source.dart:166]: Maybe improve the sentence a little...

File: test/get/hosted/resolve_with_retracted_package_versions_test.dart [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            '''Because every version of foo depends on bar ^1.0.0 which doesn't match any versions, foo is forbidden. 
+            So, because myapp depends on foo 1.0.0, version solving failed.''');
+  });
+
+  /// Currently retraction does not affect prioritization. I.e., if
+  /// pubspec.lock already contains a retracted version, which is the newest
+  /// satisfying the dependency contstraint we will not choose to downgrade.
+  /// In this case we expect a newer version to be published at some point which
+  /// will then cause pub upgrade to choose that one.
+  test('Allow retracted version when it was already in pubspec.lock', () async {
```
Comment (jonasfj) [inline test/get/hosted/resolve_with_retracted_package_versions_test.dart:51]: nit: `///` for dartdoc documentation comments, maybe just use `//`

**Review by jonasfj — APPROVED** [review]

File: lib/src/source.dart [hunk 6]
```diff
@@ -161,15 +161,20 @@ abstract class BoundSource {
...
+      {Duration maxAge, Version allowedRetractedVersion}) {
     if (ref.isRoot) {
       throw ArgumentError('Cannot get versions for the root package.');
     }
     if (ref.source != source) {
       throw ArgumentError('Package $ref does not use source ${source.name}.');
     }
 
-    return doGetVersions(ref, maxAge);
+    return doGetVersions(ref, maxAge, allowedRetractedVersion);
```
Comment (sigurdm) [inline lib/src/source.dart:177]: If possible we should do the filtering here instead of in each concrete source.

File: lib/src/source.dart [hunk 7]
```diff
@@ -161,15 +161,20 @@ abstract class BoundSource {
   /// Sources should not override this. Instead, they implement [doGetVersions].
   ///
   /// If [maxAge] is given answers can be taken from cache - up to that age old.
-  Future<List<PackageId>> getVersions(PackageRef ref, {Duration maxAge}) {
+  ///
+  /// The [allowedRetractedVersion], if given, is the only version which can be
+  /// selected even if it is marked as retracted. Otherwise, all the returned
+  /// IDs correspond to non-retracted versions.
+  Future<List<PackageId>> getVersions(PackageRef ref,
+      {Duration maxAge, Version allowedRetractedVersion}) {
```
Comment (sigurdm) [inline lib/src/source.dart:169]: Maybe this should be 
```suggestion
      {Duration maxAge, bool allowedRetractedVersions = false}) {
```
To not let the solver logic affect the version listing

File: lib/src/solver/package_lister.dart [hunk 8]
```diff
@@ -36,6 +36,15 @@ class PackageLister {
...
   final PackageId _locked;
 
+  // The version of this package that, if retracted, is still allowed in the
+  // current version solve.
+  //
+  // We don't allow a retracted version during solving unless it is already
+  // present in `pubspec.lock` or pinned in `dependency_overrides`.
+  //
+  // This is `null` if there is no retracted version that can be allowed.
+  final Version allowedRetractedVersion;
```
Comment (sigurdm) [inline lib/src/solver/package_lister.dart:46]: I have a feeling we might want `_locked` and `allowedRetractedVersion` should be the same type.

File: lib/src/solver/version_solver.dart [hunk 9]
```diff
@@ -431,7 +431,9 @@ class VersionSolver {
       List<PackageId> packages) async {
     var availableVersions = <String, List<Version>>{};
     for (var package in packages) {
-      var cached = _packageListers[package.toRef()]?.cachedVersions;
+      var packageLister = _packageListers[package.toRef()];
```
Comment (sigurdm) [inline lib/src/solver/version_solver.dart:434]: I get a feeling we should never ask the packageLister here.

The sources will cache the available versions themselves.

This might be for another PR...

File: lib/src/solver/package_lister.dart (hunk 8 above)
Comment (sigurdm) [inline lib/src/solver/package_lister.dart:46]: Can this be private?

File: lib/src/source.dart (hunk 7 above)
Comment (szakarias) [reply lib/src/source.dart:169]: As we discussed, we will not do this, but keep the `Version allowedRetractedVersion` argument.

File: lib/src/solver/package_lister.dart (hunk 8 above)
Comment (szakarias) [reply lib/src/solver/package_lister.dart:46]: Done

Comment (szakarias) [reply lib/src/solver/package_lister.dart:46]: I have the same feeling, but I think it might be `_locked` that needs to change. I'll add to the cleanup I want to do in a separate PR.

File: lib/src/solver/version_solver.dart (hunk 9 above)
Comment (szakarias) [reply lib/src/solver/version_solver.dart:434]: Noted for another PR

File: lib/src/solver/version_solver.dart (hunk 2 above)
Comment (szakarias) [reply lib/src/solver/version_solver.dart:433]: I agree, done.

File: lib/src/source.dart [hunk 10]
```diff
@@ -161,15 +161,32 @@ abstract class BoundSource {
...
     }
     if (ref.source != source) {
       throw ArgumentError('Package $ref does not use source ${source.name}.');
     }
 
-    return doGetVersions(ref, maxAge);
+    var versions = await doGetVersions(ref, maxAge);
+
+    versions = (await Future.wait(versions.map((id) async {
+      final packageStatus = await status(id, Duration(days: 3));
```
Comment (sigurdm) [inline lib/src/source.dart:180]: Use `maxAge` here instead of 3 days.

File: lib/src/source/hosted.dart [hunk 11]
```diff
@@ -454,6 +454,7 @@ class BoundHostedSource extends CachedSource {
       versionListing ??= await _cachedVersionListingResponse(ref, maxAge);
     }
     versionListing ??= await _scheduler.schedule(ref);
+
```
Comment (sigurdm) [inline lib/src/source/hosted.dart:457]: Perhaps revert this newline...

File: test/get/hosted/resolve_with_retracted_package_versions_test.dart [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      ..serve('foo', '1.0.0', deps: {'bar': '^1.0.0'})
+      ..serve('bar', '1.0.0'));
+    await d.appDir({'foo': '1.0.0'}).create();
+
+    globalPackageServer
+        .add((builder) => builder..retractPackageVersion('bar', '1.0.0'));
+    await pubGet(
+        error:
+            '''Because every version of foo depends on bar ^1.0.0 which doesn't match any versions, foo is forbidden. 
+            So, because myapp depends on foo 1.0.0, version solving failed.''');
```
Comment (sigurdm) [inline test/get/hosted/resolve_with_retracted_package_versions_test.dart:43]: Nit: the tripple-quote string ignores leading white-space, so you can write the (IMO) slightly nicer:
```suggestion
            '''
Because every version of foo depends on bar ^1.0.0 which doesn't match any versions, foo is forbidden. 
So, because myapp depends on foo 1.0.0, version solving failed.''');
```

**Review by sigurdm — APPROVED** [review]
LGTM with a few comments

File: lib/src/source.dart (hunk 10 above)
Comment (szakarias) [reply lib/src/source.dart:180]: Of course! Done.

File: lib/src/source/hosted.dart (hunk 11 above)
Comment (szakarias) [reply lib/src/source/hosted.dart:457]: Done

**Review by sigurdm — APPROVED** [review]
LGTM
